storage/namespace.go

```diff
@@ -101,31 +128,70 @@ func newDatabaseNamespace(
 	instrumentOpts := opts.InstrumentOptions()
 	scope := instrumentOpts.MetricsScope()
 	samplingRate := instrumentOpts.MetricsSamplingRate()
-	n := &dbNamespace{
-		id:       idClone,
-		shardSet: shardSet,
-		opts:     opts,
-		nsOpts:   nsMeta.Options(),
-		metrics:  newDatabaseNamespaceMetrics(scope, samplingRate),
-		nowFn:    opts.ClockOptions().NowFn(),
-	}
-	n.initShards()
-	return n
+
+	fieldBuilderOpts := indexfield.NewDocsFieldBuilderOptions().
+		SetBoolArrayPool(opts.BoolArrayPool()).
+		SetIntArrayPool(opts.IntArrayPool()).
+		SetDoubleArrayPool(opts.DoubleArrayPool()).
+		SetStringArrayPool(opts.StringArrayPool()).
+		SetInt64ArrayPool(opts.Int64ArrayPool()).
+		// Reset string array to avoid holding onto documents after we've returned the referencing
+		// array to the memory pool.
+		SetStringArrayResetFn(strings.ResetArray)
+	segmentBuilderOpts := segment.NewBuilderOptions().
+		SetFieldBuilderOptions(fieldBuilderOpts).
+		SetTimestampFieldPath(opts.TimestampFieldPath()).
+		SetRawDocSourceFieldPath(opts.RawDocSourceFieldPath()).
+		SetFieldHashFn(opts.FieldHashFn())
+
+	segmentOpts := &segmentOptions{
+		nowFn:                opts.ClockOptions().NowFn(),
+		unloadAfterUnreadFor: opts.SegmentUnloadAfterUnreadFor(),
+		instrumentOptions:    instrumentOpts,
+		fieldRetriever:       opts.FieldRetriever(),
+		executor:             opts.QueryExecutor(),
+	}
+
+	return &dbNamespace{
+		id:                  idClone,
+		opts:                opts,
+		nsOpts:              nsMeta.Options(),
+		segmentBuilderOpts:  segmentBuilderOpts,
+		segmentOpts:         segmentOpts,
+		logger:              opts.InstrumentOptions().Logger(),
+		sealedByMaxTimeAsc:  skiplist.New(),
+		metrics:             newDatabaseNamespaceMetrics(scope, samplingRate),
+		nowFn:               opts.ClockOptions().NowFn(),
+		newSegmentBuilderFn: segment.NewBuilder,
+	}
 }
 
 func (n *dbNamespace) ID() []byte { return n.id }
 
-func (n *dbNamespace) Write(
+func (n *dbNamespace) WriteBatch(
 	ctx context.Context,
-	doc document.Document,
+	docs []document.Document,
 ) error {
-	shard, err := n.shardFor(doc.ID)
-	if err != nil {
-		return err
+	builder := n.newSegmentBuilderFn(n.segmentBuilderOpts)
+	builder.AddBatch(docs)
+	rawSegment := builder.Build()
+	newSegment := newDatabaseSegment(n.id, rawSegment, n.segmentOpts)
```

Comment:
Do we want to create a new segment for every batch? Some batches could be very small. OTOH, I'm guessing this is to make it so that we're only querying immutable segments?

Is the intention that we move immediately to adding compaction to the flushing process?

I also don't see `MaxNumDocsPerSegment` used anywhere.

Reply:
Yes. Agreed on potential small batch sizes. We can either do more aggressive batching on the ingestion tier, or having some kind of ingestion queue in the database that batches things up on a fixed time interval. Either way this mechanism can be extended to handle that.

If ingestion teir handles batching logs into large batches, we don't need compaction immediately. Also I'd be curious to see what the query perf looks like with no compaction. But intuitively I think we'll need compaction at some point.

Not sure why linter doesn't complain about `MaxNumDocsPerSegment`. I'll remove.